lib/facebook/messenger/configuration/providers/base.rb

```diff
@@ -7,6 +9,18 @@ module Providers
         #   Be sure to implement all the functions as it raises
         #   NotImplementedError errors.
         class Base
+          include Facebook::Messenger::Configuration::Helpers
+
+          # A default caching implentation of generating the app_secret_proof
+          # for a given page_id
+          def app_secret_proof_for(page_id = nil)
+            app_secret = app_secret_for(page_id)
+            access_token = access_token_for(page_id)
+            @cached_app_secret_proof ||= {}
+            @cached_app_secret_proof[[app_secret, access_token]] =
```

Comment:
For me it does not look like it really caches calculated value. 
It calculates new value each time method is called. It just ensures the instance variable @cached_app_secret_proof is defined with default empty hash value.

Personally, i don't like the naming with cache cause it may potentially confuse someone that it's been 'really' cached in some cache storage. Here it is more memoization.

App secret proof usage should not be used by default (If its not required by API, for sure), so to use it you need to explicitly enable it by using ENV variable like you already suggested or by setting the flag on configuration provider instance when you build it.

The implementation i would suggest, including the ability for people to opt out from using the proof, is the following (combining with your changes):

Inside Configuration::Base:
```suggestion
attr_accessor :app_secret_proof_enabled

def initialize(*)
  self.app_secret_proof_enabled = fetch_app_secret_proof_enabled
end
  
def app_secret_proof_for(page_id = nil)
  return unless app_secret_proof_enabled

  memo_key = [app_secret_for(page_id), access_token_for(page_id)]
  memoized_app_secret_proofs[memo_key] ||= calculate_app_secret_proof(app_secret, access_token)
end

private

def memoized_app_secret_proofs
  @memoized_app_secret_proofs ||= {}
end

def fetch_app_secret_proof_enabled
  false
end
```

Inside Configuration::Environment:
```suggestion
              def fetch_app_secret_proof_enabled
    ENV['APP_SECRET_PROOF_ENABLED']
  end
```

Reply:
@egor-moysevich looks a little bit cleaner than what I had, and includes the suggested improvement for `fetch_app_secret_proof_enabled`, so I'll integrate the changes.
Agree with you about the naming of the variable memoisation seems clearer

> It calculates new value each time method is called. It just ensures the instance variable @cached_app_secret_proof is defined with default empty hash value.
You're right, sorry, that was a typo that I introduced while testing the caching (the second line should be `||=` as well, I think I turned the caching off so I didn't have to keep restarting during testing, and didn't turn it back).

I'll integrate the changes and put in a test that `calculate_app_secret_proof` only gets hit once upon repeated calls to the function.